File: src/main/java/de/byjoker/myjfql/command/CommandServiceImpl.java
package de.byjoker.myjfql.command;

import de.byjoker.myjfql.exception.LanguageException;
import de.byjoker.myjfql.lang.CommandFormatter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.google.common.reflect.ClassPath.ClassInfo;
import static com.google.common.reflect.ClassPath.from;

public class CommandServiceImpl implements CommandService {

    private final List<Command> commands;
    private final CommandFormatter formatter;

    public CommandServiceImpl(CommandFormatter formatter) {
        this.commands = new ArrayList<>();
        this.formatter = formatter;
    }

    @Override
    public void registerCommand(Command command) {
        commands.add(command);
    }

    @Override
    public void unregisterCommand(Command command) {
        commands.remove(command);
    }

    @Override
    public void searchCommands(String path) {
        final ClassLoader loader = Thread.currentThread().getContextClassLoader();

        try {
            for (ClassInfo info : from(loader).getTopLevelClasses()) {
                if (info.getName().startsWith(path)) {
                    try {
                        final Class<? extends Command> clazz = (Class<? extends Command>) info.load();

                        if (clazz.isAnnotationPresent(CommandHandler.class))
                            registerCommand(clazz.newInstance());
                    } catch (Exception ignore) {
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean existsCommand(String name) {
        return commands.stream().anyMatch(command -> command.getName().equalsIgnoreCase(name));
    }

    @Override
    public Command getCommand(String name) {
        return commands.stream().filter(command -> command.getName().equalsIgnoreCase(name)).findFirst().orElse(null);
    }

    @Override
    public void execute(CommandSender sender, String name) {
        try {
            final Map<String, List<String>> arguments = formatter.formatCommand(name);

            if (arguments == null) {
                sender.sendError("Command was not found!");
                return;
            }

            final Command command = getCommand(arguments.get("COMMAND").get(0));

            if (command == null) {
                sender.sendError("Command was not found!");
                return;
            }

            command.execute(sender, arguments);
        } catch (Exception ex) {
            new LanguageException(ex).printStackTrace();
        }
    }

    @Override
    public List<Command> getCommands() {
        return commands;
    }

}

File: src/main/java/de/byjoker/myjfql/command/CommandServiceImpl.kt
package de.byjoker.myjfql.command

import de.byjoker.myjfql.lang.Interpreter
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.stream.Collectors


class CommandServiceImpl(private var commands: MutableList<Command> = mutableListOf()) : CommandService {

    override fun registerCommand(command: Command) {
        commands.add(command)
    }

    override fun unregisterCommand(name: String) {
        commands.removeIf { command -> command.name.equals(name, ignoreCase = true) }
    }

    override fun searchCommands(directory: String) {
        for (clazz in findAllClassesUsingClassLoader(directory)) {
            if (clazz.isInstance(Command::class)) registerCommand(
                (clazz as Class<Command>).getDeclaredConstructor().newInstance()
            )
        }
    }

    override fun getCommand(name: String): Command? {
        return commands.firstOrNull { command -> command.name.equals(name, ignoreCase = true) }
    }

    override fun getCommands(): List<Command> = commands

    override fun execute(sender: CommandSender, query: String, interpreter: Interpreter) {
        try {
            val arguments = interpreter.interpretCommand(query)

            if (!arguments.containsKey("command")) {
                sender.error("Command wasn't defined!")
                return
            }

            val command = getCommand(arguments["command"]!![0])

            if (command == null) {
                sender.error("Command wasn't found!")
                return
            }

            command.execute(sender, arguments)
        } catch (ex: Exception) {
            sender.error(ex)
        }
    }

    private fun findAllClassesUsingClassLoader(packageName: String): List<Class<*>> {
        val stream: InputStream =
            ClassLoader.getSystemClassLoader().getResourceAsStream(packageName.replace("[.]".toRegex(), "/"))
        val reader = BufferedReader(InputStreamReader(stream))
        return reader.lines().filter { line: String -> line.endsWith(".class") }.map { line: String ->
            getClass(
                line, packageName
            )
        }.collect(Collectors.toSet()).filter { clazz -> clazz.isAnnotationPresent(CommandHandler::class.java) }
    }

    private fun getClass(className: String, packageName: String): Class<*> {
        return Class.forName(
            packageName + "." + className.substring(0, className.lastIndexOf('.'))
        )
    }

}
